crates/ruma-state-res/src/event_auth.rs

```diff
@@ -158,44 +157,65 @@ pub fn check_state_independent_auth_rules<E: Event>(
+        HashSet::with_capacity(expected_auth_types.len());
 
     // Since v1, considering auth_events:
-    //
-    // - Since v1, if there are duplicate entries for a given type and state_key pair, reject.
-    //
-    // - Since v1, if there are entries whose type and state_key don’t match those specified
-    //   by the auth events selection algorithm described in the server specification, reject.
-    let expected_auth = auth_types_for_event(
-        incoming_event.kind,
-        sender,
-        incoming_event.state_key,
-        incoming_event.content().clone(),
-    );
-
-    dbg!(&expected_auth);
-
-    for ev_key in auth_events.keys() {
-        // (b)
-        if !expected_auth.contains(ev_key) {
-            warn!("auth_events contained invalid auth event");
-            return Ok(false);
+    for auth_event_id in incoming_event.auth_events() {
+        let event_id = auth_event_id.borrow();
+
+        let Some(auth_event) = fetch_event(event_id) else {
+            return Err(format!("failed to find auth event {event_id}"));
+        };
+
+        // The auth event must be in the same room as the incoming event.
+        if auth_event.room_id() != room_id {
+            return Err(format!("auth event {event_id} not in the same room"));
         }
-    }
-    */
 
-    // TODO:
-    //
-    // Since v1, if there are entries which were themselves rejected under the checks performed on
-    // receipt of a PDU, reject.
+        let event_type = auth_event.event_type();
+        let state_key = auth_event
+            .state_key()
+            .ok_or_else(|| format!("auth event {event_id} has no `state_key`"))?;
+        let key = (event_type.clone(), state_key.to_owned());
+
+        // Since v1, if there are duplicate entries for a given type and state_key pair, reject.
+        if seen_auth_types.contains(&key) {
+            return Err(format!(
+                "duplicate auth event {event_id} for ({event_type}, {state_key}) pair"
+            ));
+        }
+
+        // Since v1, if there are entries whose type and state_key don’t match those specified by
+        // the auth events selection algorithm described in the server specification, reject.
+        if !expected_auth_types.contains(&key) {
+            return Err(format!(
+                "unexpected auth event {event_id} with ({event_type}, {state_key}) pair"
+            ));
+        };
+
+        // TODO:
+        //
+        // Since v1, if there are entries which were themselves rejected under the checks performed
+        // on receipt of a PDU, reject.
+
+        seen_auth_types.insert(key);
+    }
 
     // Since v1, if there is no m.room.create event among the entries, reject.
-    if !incoming_event.auth_events().any(|id| {
-        fetch_event(id.borrow())
-            .is_some_and(|event| *event.event_type() == TimelineEventType::RoomCreate)
-    }) {
+    if !seen_auth_types.iter().any(|(event_type, _)| *event_type == TimelineEventType::RoomCreate) {
```

Comment:
```suggestion
    if !seen_auth_types.contains(&(TimelineEventType::RoomCreate, "".to_string)) {
```

Assuming that we're actually guaranteed to never see `m.room.create` with a nonempty `state_key`, which I'm [not 100% certain about](https://matrix.to/#/#matrix-spec:matrix.org/$Sq9NgJq3tS5aIid2ZT_FLXQvNMIJt9xWQVkLuW5tDpw).

Reply:
That's what I did previously, but I preferred to keep the same behavior as Synapse to avoid surprises. I don't have strong feelings either way so if you prefer to use `contains`, I'll change the code.

Reply:
I think keeping the same behavior as synapse is probably the way to go here.